enclave-runtime/src/rpc/worker_api_direct.rs

```diff
@@ -89,33 +88,73 @@ fn compute_encoded_return_error(error_msg: String) -> Vec<u8> {
+where
+	R: AuthorApi<H256, H256>
+		+ SendState<Hash = H256>
+		+ OnBlockCreated<Hash = H256>
+		+ Send
+		+ Sync
+		+ 'static,
+{
 	let mut io = IoHandler::new();
 	let mut rpc_methods_vec: Vec<&str> = Vec::new();
 
 	// Add rpc methods
 	// author_submitAndWatchExtrinsic
 	let author_submit_and_watch_extrinsic_name: &str = "author_submitAndWatchExtrinsic";
-	let submit_watch_pool_getter = tx_pool_getter.clone();
+	let submit_watch_author = rpc_author.clone();
 	rpc_methods_vec.push(author_submit_and_watch_extrinsic_name);
 	io.add_sync_method(author_submit_and_watch_extrinsic_name, move |params: Params| match params
 		.parse::<Vec<u8>>()
 	{
-		Ok(encoded_params) => {
-			let tx_pool_guard = submit_watch_pool_getter.get().unwrap().lock().unwrap();
-			let tx_pool = Arc::new(tx_pool_guard.deref());
-			let author = Author::new(tx_pool);
+		Ok(encoded_params) => match Request::decode(&mut encoded_params.as_slice()) {
+			Ok(request) => {
+				let shard: ShardIdentifier = request.shard;
+				let encrypted_trusted_call: Vec<u8> = request.cyphertext;
+				let result = async {
+					submit_watch_author.watch_top(encrypted_trusted_call.clone(), shard).await
+				};
+				let response: Result<Hash, RpcError> = executor::block_on(result);
+				let json_value = match response {
+					Ok(hash_value) => RpcReturnValue {
+						do_watch: true,
+						value: hash_value.encode(),
+						status: DirectRequestStatus::TrustedOperationStatus(
+							TrustedOperationStatus::Submitted,
+						),
+					}
+					.encode(),
+					Err(rpc_error) => compute_encoded_return_error(rpc_error.message),
+				};
+				Ok(json!(json_value))
+			},
+			Err(_) =>
+				Ok(json!(compute_encoded_return_error("Could not decode request".to_owned()))),
+		},
+		Err(e) => {
+			let error_msg: String = format!("Could not submit trusted call due to: {}", e);
+			Ok(json!(compute_encoded_return_error(error_msg)))
+		},
+	});
 
-			match Request::decode(&mut encoded_params.as_slice()) {
+	// author_submitExtrinsic
+	let author_submit_extrinsic_name: &str = "author_submitExtrinsic";
+	let submit_author = rpc_author.clone();
+	rpc_methods_vec.push(author_submit_extrinsic_name);
+	io.add_sync_method(author_submit_extrinsic_name, move |params: Params| {
+		match params.parse::<Vec<u8>>() {
+			Ok(encoded_params) => match Request::decode(&mut encoded_params.as_slice()) {
 				Ok(request) => {
 					let shard: ShardIdentifier = request.shard;
-					let encrypted_trusted_call: Vec<u8> = request.cyphertext;
-					let result =
-						async { author.watch_top(encrypted_trusted_call.clone(), shard).await };
+					let encrypted_trusted_op: Vec<u8> = request.cyphertext;
+					let result = async {
+						submit_author.submit_top(encrypted_trusted_op.clone(), shard).await
+					};
 					let response: Result<Hash, RpcError> = executor::block_on(result);
 					let json_value = match response {
 						Ok(hash_value) => RpcReturnValue {
-							do_watch: true,
+							do_watch: false,
```

Comment:
Was this a bug before?

Reply:
This tripped me too the first time I looked at it. No, this is just the diff being confusing, if you look at the whole change set, you will see that the diff matched two different parts of the code here.